8 card 도메인 구현

sillysillyman opened the pull request:
Issue: #8 
Card 도메인 CRUD 구현 및 커스텀 예외 처리 추가

sillysillyman committed:
[Feat]: CardAdapter 클래스 추가

- CardRepository를 의존성 주입 받아 사용하기 위한 CardAdapter 클래스 추가

sillysillyman committed:
[Feat]: CardController 클래스 추가

- CardService를 의존성 주입 받아 사용하기 위한 CardController 클래스 추가

sillysillyman committed:
[Feat]: CardRepository 인터페이스 추가

- Card 엔티티를 위한 JPA 리포지토리 인터페이스 추가

sillysillyman committed:
[Feat]: CardService 클래스 추가

- CardAdapter를 의존성 주입 받아 Card 엔티티의 비즈니스 로직을 처리하기 위한 CardService 클래스 추가

sillysillyman committed:
Merge branch 'dev' of https://github.com/TrelloProject/Trello into 8-card-도메인-구현

sillysillyman committed:
[Feat]: Card 엔티티 수정

- Card 엔티티에 Builder 패턴 추가
- 필드에 @Setter 어노테이션 추가

sillysillyman committed:
[Refactor]: 클래스 이동

sillysillyman committed:
[Feat]: CreateCardRequestDto 추가

- CreateCardRequestDto 클래스 추가
- 덱 ID, 카드 제목, 카드 설명 필드 추가 및 유효성 검사 애너테이션 적용

sillysillyman committed:
[Feat]: MoveCardRequestDto 추가

- MoveCardRequestDto 클래스 추가
- 덱 ID 및 인덱스 필드 추가

sillysillyman committed:
[Feat]: UpdateCardRequestDto 추가

- UpdateCardRequestDto 클래스 추가
- title, description, startDate, dueDate 필드 추가
- LocalDate 필드에 대한 JsonFormat 애너테이션 추가

sillysillyman committed:
[Feat]: CardRepository 메서드 추가

- 카드 이동 및 검색을 위한 메서드 추가
  - findCardByNextId(Long cardId)
  - findTopByDeckIdOrderByIdDesc(Long deckId)
  - findAllByDeckId(Long deckId)

sillysillyman committed:
[Feat]: CardAdapter 기능 추가

- 카드 저장, 삭제, 조회 및 소유권 확인 메서드 추가

sillysillyman committed:
[Feat]: 카드 서비스 기능 구현

- 카드 생성 기능 추가: 덱의 헤드 카드 설정 로직 포함
- 카드 조회 기능 추가: 사용자 소유권 확인 포함
- 카드 업데이트 기능 추가: 카드 제목, 설명, 시작일, 마감일 업데이트
- 카드 이동 기능 추가: 덱 내에서의 카드 위치 변경 로직 구현
- 카드 삭제 기능 추가: 덱의 헤드 카드 갱신 로직 포함

sillysillyman committed:
[Feat]: 카드 컨트롤러 기능 구현

- 카드 생성 엔드포인트 추가: `/cards` (POST)
- 카드 조회 엔드포인트 추가: `/cards/{cardId}` (GET)
- 카드 업데이트 엔드포인트 추가: `/cards/{cardId}` (PUT)
- 카드 이동 엔드포인트 추가: `/cards/{cardId}` (PATCH)
- 카드 삭제 엔드포인트 추가: `/cards/{cardId}` (DELETE)

sillysillyman committed:
[Refactor]: 패키지 경로 수정

- CardAdapter의 패키지 경로를 repository에서 adapter로 수정

sillysillyman committed:
[Refactor]: Deck 엔티티 수정

- headCardId 필드에 @Setter 추가
- 구글 스타일 가이드 적용

sillysillyman committed:
[Feat]: Card 도메인 api 테스트 추가

sillysillyman committed:
[Conflict]: merge 후 충돌 해결

sillysillyman committed:
[Refactor]: CardAdapter 커스텀 예외 처리 및 카드 정렬 로직 추가

- DeckRepository 의존성 추가
- 카드 소유권 검증 시 커스텀 예외 처리 추가 (CardNoPermissionException)
- 카드 조회 시 커스텀 예외 처리 추가 (CardNotFoundException)
- findCardByDeckIdAndIndex 메서드에 덱의 헤드 카드 및 정렬 로직 추가
- 덱이 비어 있는 경우와 인덱스가 범위를 벗어나는 경우 예외 처리 추가

sillysillyman committed:
[Feat]: CardCodeEnum 추가

- 카드 관련 예외 코드를 관리하는 CardCodeEnum 추가
- 카드가 존재하지 않을 때의 예외 코드(CARD_NOT_FOUND) 추가
- 카드에 대한 권한이 없을 때의 예외 코드(CARD_NO_PERMISSION) 추가

sillysillyman committed:
[Feat]: CardException 클래스 추가

- CardException 클래스 추가
- 카드 관련 예외 처리를 위한 CardCodeEnum 필드 추가
- CardCodeEnum을 이용해 예외 메시지 설정

sillysillyman committed:
[Feat]: CardNoPermissionException 클래스 추가

- CardNoPermissionException 클래스 추가
- CardException 클래스를 상속받아 카드 권한 관련 예외 처리
- CardCodeEnum을 이용해 예외 메시지 설정

sillysillyman committed:
[Feat]: CardNotFoundException 클래스 추가

- CardNotFoundException 클래스 추가
- CardException 클래스를 상속받아 카드 조회 관련 예외 처리
- CardCodeEnum을 이용해 예외 메시지 설정

sillysillyman committed:
[Feat]: CreateCardRequestDto에 startDate 및 dueDate 필드 추가

- CreateCardRequestDto에 startDate 및 dueDate 필드 추가
- JsonFormat 애너테이션을 사용하여 날짜 형식 지정

sillysillyman committed:
[Feat]: GlobalExceptionAdvice에 CardException 처리 추가

- GlobalExceptionAdvice에 CardException 핸들러 추가
- CardException 발생 시 적절한 HTTP 상태 코드와 메시지 반환

sillysillyman committed:
[Refactor]: CardService 클래스 리팩토링

- createCard 메서드에서 startDate와 dueDate 필드 처리 로직 추가
- moveCard 메서드에서 이동할 카드의 앞 뒤를 연결하는 로직 추가

sillysillyman committed:
[Feat]: Card 엔티티 인덱스 생성

- deck_id 인덱스 생성
- next_id 인덱스 생성
- title, description, start_date, due_date 복합 인덱스 생성

sillysillyman committed:
[Feat]: Fetch Join 메서드 구현

- Card 엔티티와 연관관계에 있는 엔티티에 대해 Fetch Join 메서드 구현

sillysillyman committed:
Merge branch 'dev' of https://github.com/TrelloProject/Trello into 8-card-도메인-구현

tjdrbs0712 commented on src/main/java/com/sparta/trello/domain/card/adapter/CardAdapter.java:
```diff
@@ -0,0 +1,90 @@
+        List<Card> cards = findAllByDeckId(deckId);
+
+        if (cards.isEmpty()) {
+            throw new IllegalArgumentException("덱이 비어 있습니다."); // 커스텀 예외 추가 예정
```
예외 처리 추가 해주셔야 될거 같습니다.

sillysillyman replied:
추가했습니다

sillysillyman committed:
[Feat]: CardCodeEnum 수정

- CARD_INDEX_OUT_OF_BOUNDS_ERROR 항목 추가

sillysillyman committed:
[Feat]: CardIndexOutOfBoundsException 추가

- 카드 인덱스 범위 초과 예외 처리 클래스 추가

sillysillyman committed:
[Feat]: CardAdapter에 커스텀 예외 처리 추가 및 적용

- CardIndexOutOfBoundsException 추가 및 findCardByDeckIdAndIndex 메서드에서 적용
- 빈 덱인 경우 던져지는 예외를 DeckDetailCustomException으로 수정

sillysillyman committed:
[Feat]: Card 엔티티에 update 메서드 추가 및 Setter 제거

- UpdateCardRequestDto를 사용하여 엔티티 필드를 업데이트하는 update 메서드 추가
- 필드 업데이트를 위해 title, description, startDate, dueDate의 Setter 제거

sillysillyman committed:
[Fix]: CreateCommentRequestDto 카드 ID 유효성 검사 수정

- 카드 ID 필드에 @NotNull 애너테이션 추가하여 유효성 검사 수정

sillysillyman committed:
[Refactor]: updateCard 코드 개선

- 카드 업데이트 시 Card 엔티티의 update 메서드를 통해 필드 업데이트
